### DISCUSSION TITLE
How to make the Awaited type pass the error test case?

### DISCUSSION THREAD

**Discussion by mrdulin** [type-challenges/type-challenges#22408] [Q&A]
It's so weird. I can't think of any way. 

```ts
type MyAwaited<T> = T extends PromiseLike<infer V> ? MyAwaited<V> : T

// @ts-expect-error
type error = MyAwaited<number>
```
Got error: 
> Unused '@ts-expect-error' directive.

So I think I didn't pass the error test case.

[TS Playground](https://www.typescriptlang.org/play?#code/PQKgUABBCMAcCcEC0ECCB3AhgSwC4FMATSZJM8kgIwE8IBZTAY23wCsIBlbAawHsAnTBAAUAAQC2TFqwDOPAZgCUEAMT5MM2ioAO-XuOwz8qygFdsAG1xJsAOzAkVTiAEVT+Gbmy97JAJIAZhDoxgAWmABuxkK41NrG6KHYjKEQhsGC2vGEELHxEBY8xgAKegZGAHQQABK86MHGjJi2EADm+LgQMXEJSSlpMmm2coTGuKEJmdm5PQD8DlAAYgIQ+AAemOLaFvgAXGlBIRDhURAABqX6hvgAPACiG1s7ACo9AHxnx3W5vG0dEA9Ntt8K94vMSGdIbgZCQ8sZAU8QT0IABeCCXcq3Tz8OytN4LGb5ABKHlMVlR9GoGBwBEI90ewNB+DeEGAwAg2NxEMhBJZzySgwAju5PN4WultAJaRAAmVchMIABtATYVp2TAWLr8LyMHYAXWEoVwuG0Ml2bNGEQquF4wEkjDkfEEwEwhAizUYRCQcJkjBx2ms63w-GYRhkSGFHi8PiQ0CQABYAKzcePKGhKiRSNiOhQGo0ms1stXjUyUCqMfR2rOyeSCRS8iAANRY9R8EAA4nhqqX9vnTebgNCUhVZBUBK1gHB4GAQMAHKAIAB9Zcr1criAATV4pn4EAAwrxRjVg8Y12flxBZw44ZTqXgiDdniy0c9VmsCLZCIMMdcADJFG47ACYMmxZWZbywe86UbFl9meecQCXc81wgZ4o33DQPCQ5D1yvbAtilQljAAbwBYUNQAGgBNZ4kYToAF8ZTlAByUQ4SQFINR2Wx2hkYBTC8CwZGY69kQADQpH8jBuTkePxG8N0ksprhuUiAhYCxCH2WxTHESgQPo+TkQALSUq5pKkrFcBxHiIAAHwgHS9ODN4jPyYzoDMzEbksnzlOk2TWnsiBKF4XgdmaVy3OMV80VI8Z8FsfZhB8AIyXUiwdi0kRMH4VptN0-T+GUFEWWaagSrK2xaHo0T8iaMMKUVEgHlo3B7nIiwbjoKlINpG4xLeKjAtciiWpo-A6I60wNW63qaQfDchogNSNOypyiogQyhvGtrptmnq7364zlsC4KNpcnaoFayb2ruTq5qOh8PNO6zcWC0LwvUWxRt2279q6w6+ofJ8qIu-hfr1ed2TY8N1japBgz0fgwBvJGVjRIGFrpcH8TABccNXCBFh3BLdw4AhTWwwnLznUASBZDhwn4YxqG3XcZHCgSxTNY5jX7Nkh1CEcZDHPLJwQF1hhCFGoBZZt8HqTmLG5nxeb7QtB19YXR3HCX4GAZXVeGBn6AEYw93CTLEt43t+c1oWRbF1oZznIA)

**Comment by ChrisKulesza** [comment]
@mrdulin 
You must restrict your generic type of MyAwaited to PromiseLike<unknown>. Like so. But then you still haven't reached your destination. Since you still have to take care of nested PromiseLikes.

```ts
type MyAwaited<T extends PromiseLike<unknown>> = T extends PromiseLike<infer V> ? MyAwaited<V> : T
```

**Reply by mrdulin** [reply]
Thanks for your reply. But TSC throws an error for your answer:

> Type 'V' does not satisfy the constraint 'PromiseLike<unknown>'.(2344)

**Reply by jazelly** [reply]
> Type 'V' does not satisfy the constraint 'PromiseLike'.(2344)

This error simply means V does not satisfy the `T extends PromiseLike<unknown>` condition. To suppress it, You can simply add check after infer, e..g `infer V extends PromiseLike<any>`, or defer it to next if else

**Comment by ChrisKulesza** [comment]
I know. That's why I gave you the recursion hint. If this is not enough for you, then you can also work with the solutions of the other users here.
